src/vma/sock/sockinfo_udp.cpp

```diff
@@ -939,13 +941,20 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
 			case IP_MULTICAST_TTL:
 				{
 					int n_mc_ttl = -1;
-					if (__optlen == 1)
+					if (__optlen == sizeof(m_n_mc_ttl))
 						n_mc_ttl = *(char*)__optval;
-					else if (__optlen >= 1)
+					else if (__optlen == sizeof(int))
 						n_mc_ttl = *(int*)__optval;
-
-					if (__optval && (n_mc_ttl >= 0 && n_mc_ttl <= 255)) {
+					else {
+						break;
+					}
+					if (n_mc_ttl == -1) {
+						n_mc_ttl = 1;
+					}
+					if (n_mc_ttl >= 0 && n_mc_ttl <= 255) {
 						m_n_mc_ttl = n_mc_ttl;
+						header_tos_updater du(m_n_mc_ttl);
```

Comment:
header_tos_updater type is wrong...
And i think we should add another type for MC.. udp socket might have diffent mc/uc ttl's ..